### ISSUE TITLE
Salmon problems with Friendica

### ISSUE THREAD

**Issue by annando** [tootsuite/mastodon#2196]
One of our users reported this issue: https://github.com/friendica/friendica/issues/3360

I tested it and can confirm this issue. Mastodon always confirms every salmon with HTTP code "201" (instead of GNU Social, which returns "200" in case of success or "400" in case of an error).

The salmon communication between Friendica and GNU Social is working fine, so it seems as if Mastodon is behaving different in some point. (Receiving of Mastodon salmons seems to work in Friendica)

Question is if there is some logging in Mastodon that could help in debugging this. Or is there some description how Mastodon verifies salmons?

**Comment by Gargron** [comment]
Ah, I see. At some point we started processing Salmon payloads in the background, so verification doesn't happen where we can return a 400: https://github.com/tootsuite/mastodon/blob/master/app/controllers/api/salmon_controller.rb#L13

It should be simple enough to move verification back out from the background job.

**Comment by annando** [comment]
Background processing itself isn't that bad. Especially since you possibly have to do a webfinger request to fetch the sender key - but surely it makes debugging much worse :-)

When you would switch that, it would of course help my debugging. If I then would get a "200" then I knew that my message was delivered successfully. 

And if you would be a fairy and I had some wishes for free then I would like to have some different "4xx" messages that could return why it failed (like "bad signature", "couldn't fetch public key", "message could be decrypted, but was rejected, ...") :-)

**Comment by annando** [comment]
I just did some tests. I changed some parts in our XML that we are generating. With this I was able to send a Salmon that was received on Mastodon as well: https://mastodon.social/@heluecht/4874165 (See the fish at the end of the post that I used to separate it from the feed transport that used a baby bottle as symbol)

Users still are reporting that Salmon doesn't work. I tested it with a Friendica user the Mastodon user is following and a Friendica user on a different server whom the Mastodon user isn't following. Are these settings responsible for it?
```
[ ] Block notifications from non-followers
[ ] Block notifications from people you don't follow
```
Or is there some additional check in the background? Can server admins apply some restrictions?
